torch/csrc/distributed/c10d/python_comm_hook.cpp

```diff
@@ -39,22 +39,21 @@ c10::intrusive_ptr<c10::ivalue::Future> PythonCommHook::runHook(
   }
 }
 
-std::vector<at::Tensor> PythonCommHook::parseHookResult(
-    const c10::IValue& result) {
+at::Tensor PythonCommHook::parseHookResult(const c10::IValue& result) {
   TORCH_INTERNAL_ASSERT(
       result.isPyObject() || result.isTensorList(),
       "expected the hook result is either a PyObject or TensorList");
 
   if (result.isPyObject()) {
     py::gil_scoped_acquire ag;
     py::object obj = torch::jit::toPyObject(result);
-    auto value = torch::jit::toIValue(
-        obj, c10::ListType::create(c10::TensorType::get()));
-
-    return value.toTensorVector();
+    auto value = torch::jit::toIValue(obj, c10::TensorType::get());
+    return value.toTensor();
   }
 
-  return result.toTensorVector();
+  // Only if the hook is an `allreduce_hook`, as the vanilla allreduce returns a
+  // tensor vector.
+  return result.toTensorVector()[0];
```

Comment:
Should we modify allreduce hook to return a single tensor?

Reply:
I see that `get_future()` returns a future[list[tensor]], is it worth it to refactor the hook to do something like

```
return dist.all_reduce(tensor).get_future().then(lambda li: return lu[0])
```

Reply:
I am thinking of `dist.all_reduce(tensor).get_future().wait()[0]` instead. @lw Could you please confirm that this will still remain the asynchronous property?

Reply:
IIUC `return dist.all_reduce(tensor).get_future().wait()[0]` won't work because it is a single tensor, but the contract for the comm. hook is that it needs to return a future to a single tensor?

Reply:
Will update the code with `dist.all_reduce(tensor).get_future().then(lambda li: return lu[0])` in another PR.